File: web/src/engine/adaptation/readinessEvaluator.ts
// Portage de Adaptation/ReadinessEvaluator.swift.
import type { DailyReadiness } from '../domain/activity';

export type ReadinessLevel = 'good' | 'moderate' | 'low';

export interface ReadinessAssessment {
  level: ReadinessLevel;
  score: number; // 0–100
  reasons: string[];
}

function standardDeviation(xs: number[], mean: number): number {
  if (xs.length <= 1) return 0;
  const variance = xs.reduce((s, v) => s + (v - mean) * (v - mean), 0) / (xs.length - 1);
  return Math.sqrt(variance);
}

export class ReadinessEvaluator {
  assess(today: DailyReadiness, history: DailyReadiness[]): ReadinessAssessment {
    let score = 100.0;
    const reasons: string[] = [];

    // HRV vs ligne de base (moyenne − écart-type)
    const hrvs = history.map((h) => h.hrvMs).filter((x): x is number => x != null);
    if (today.hrvMs != null && hrvs.length >= 3) {
      const mean = hrvs.reduce((s, v) => s + v, 0) / hrvs.length;
      const sd = standardDeviation(hrvs, mean);
      if (today.hrvMs < mean - sd) {
        score -= 30;
        reasons.push('VFC nettement sous ta normale : système nerveux fatigué.');
      } else if (today.hrvMs < mean - 0.5 * sd) {
        score -= 12;
        reasons.push('VFC légèrement basse.');
      }
    }

    // FC de repos élevée
    const rests = history.map((h) => h.hrRest).filter((x): x is number => x != null);
    if (today.hrRest != null && rests.length >= 3) {
      const mean = rests.reduce((s, v) => s + v, 0) / rests.length;
      if (today.hrRest > mean + 5) {
        score -= 15;
        reasons.push(`FC de repos élevée (+${Math.trunc(today.hrRest - mean)} bpm).`);
      }
    }

    // Sommeil — malus PROGRESSIF (une nuit très courte doit peser lourd).
    if (today.sleepHours != null) {
      const h = today.sleepHours;
      let pen = 0;
      if (h < 3) pen = 55;
      else if (h < 4) pen = 45;
      else if (h < 5) pen = 33;
      else if (h < 6) pen = 22;
      else if (h < 7) pen = 10;
      else if (h < 7.5) pen = 4;
      if (pen > 0) {
        score -= pen;
        reasons.push(`Nuit ${h < 5 ? 'très ' : ''}courte (${h.toFixed(1)} h).`);
      }
    }

    // Ressenti subjectif
    if (today.subjective) {
      const s = today.subjective;
      if (s.soreness <= 2) { score -= 15; reasons.push('Courbatures marquées.'); }
      if (s.form <= 2) { score -= 12; reasons.push('Forme ressentie basse.'); }
      if (s.motivation <= 2) { score -= 6; reasons.push('Motivation en berne.'); }
    }

    // Body Battery
    if (today.bodyBattery != null && today.bodyBattery < 30) {
      score -= 10;
      reasons.push(`Réserves d'énergie basses (Body Battery ${today.bodyBattery}).`);
    }

    score = Math.max(0, Math.min(100, score));
    let level: ReadinessLevel = score >= 70 ? 'good' : score >= 45 ? 'moderate' : 'low';

    // Garde-fous sécurité : un facteur critique ne peut jamais donner « au top ».
    const s = today.subjective;
    const veryLowSleep = today.sleepHours != null && today.sleepHours < 4;
    const veryBadFeel = !!s && (s.form <= 1 || s.soreness <= 1);
    if ((veryLowSleep || veryBadFeel) && level === 'good') level = 'moderate';
    // Nuit quasi blanche ou ressenti + courbatures au plus bas → repos impératif.
    if ((today.sleepHours != null && today.sleepHours < 3) || (!!s && s.form <= 1 && s.soreness <= 1)) {
      level = 'low';
    }

    if (reasons.length === 0) reasons.push('Bonne récupération, prêt à performer.');
    return { level, score, reasons };
  }
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Adaptation/ReadinessEvaluator.swift
import Foundation

public enum ReadinessLevel: String, Sendable, Codable {
    case good, moderate, low
}

public struct ReadinessAssessment: Sendable, Equatable {
    public var level: ReadinessLevel
    public var score: Double            // 0–100
    public var reasons: [String]        // explications pédagogiques
}

/// Évalue l'état de récupération du jour à partir des données objectives
/// (HRV, FC repos, sommeil) comparées à une ligne de base, et du ressenti.
public struct ReadinessEvaluator: Sendable {
    public init() {}

    /// `history` : jours précédents pour établir la ligne de base HRV / FC repos.
    public func assess(today: DailyReadiness, history: [DailyReadiness]) -> ReadinessAssessment {
        var score = 100.0
        var reasons: [String] = []

        // --- HRV vs ligne de base (moyenne − écart-type) ---
        let hrvs = history.compactMap(\.hrvMs)
        if let hrv = today.hrvMs, hrvs.count >= 3 {
            let mean = hrvs.reduce(0, +) / Double(hrvs.count)
            let sd = standardDeviation(hrvs, mean: mean)
            if hrv < mean - sd {
                score -= 30; reasons.append("VFC nettement sous ta normale : système nerveux fatigué.")
            } else if hrv < mean - 0.5 * sd {
                score -= 12; reasons.append("VFC légèrement basse.")
            }
        }

        // --- FC de repos élevée ---
        let rests = history.compactMap(\.hrRest)
        if let rest = today.hrRest, rests.count >= 3 {
            let mean = Double(rests.reduce(0, +)) / Double(rests.count)
            if Double(rest) > mean + 5 {
                score -= 15; reasons.append("FC de repos élevée (+\(Int(Double(rest) - mean)) bpm).")
            }
        }

        // --- Sommeil ---
        if let sleep = today.sleepHours {
            if sleep < 6 { score -= 20; reasons.append("Nuit courte (\(String(format: "%.1f", sleep)) h).") }
            else if sleep < 7 { score -= 10; reasons.append("Sommeil un peu juste.") }
        }

        // --- Ressenti subjectif (1–5, 5 = au top / aucune courbature) ---
        if let s = today.subjective {
            if s.soreness <= 2 { score -= 15; reasons.append("Courbatures marquées.") }
            if s.form <= 2 { score -= 12; reasons.append("Forme ressentie basse.") }
            if s.motivation <= 2 { score -= 6; reasons.append("Motivation en berne.") }
        }

        // --- Body Battery (si dispo via Garmin/Health) ---
        if let bb = today.bodyBattery, bb < 30 {
            score -= 10; reasons.append("Réserves d'énergie basses (Body Battery \(bb)).")
        }

        score = max(0, min(100, score))
        let level: ReadinessLevel = score >= 70 ? .good : (score >= 45 ? .moderate : .low)
        if reasons.isEmpty { reasons.append("Bonne récupération, prêt à performer.") }
        return ReadinessAssessment(level: level, score: score, reasons: reasons)
    }

    private func standardDeviation(_ xs: [Double], mean: Double) -> Double {
        guard xs.count > 1 else { return 0 }
        let variance = xs.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(xs.count - 1)
        return variance.squareRoot()
    }
}
